File: engine/lib/repo_layout.py
"""Resolve repo root and engine paths — HQ (engine/ subtree) vs user fork (flat or engine/)."""
from __future__ import annotations

from pathlib import Path


def repo_root_from_here(file: str | Path) -> Path:
    """engine/scripts/foo.py → repo root; scripts/foo.py (skeleton) → repo root."""
    here = Path(file).resolve().parent
    dir = here
    for _ in range(8):
        # HQ monorepo
        if (dir / "platform" / "soul").is_dir() and (dir / "ui").is_dir():
            return dir
        if (dir / "engine" / "soul").is_dir():
            return dir
        # Athlete repo: a composed SOUL at the propagated/ or repo root. Both the post-ADR-0022
        # `SOUL.claude.md` name and the pre-split `SOUL.md` count - repos carved before the
        # split still carry the old name and must keep resolving.
        has_athlete_band = (dir / "user_data").is_dir() or (dir / "training").is_dir()
        if (
            (dir / "propagated" / "SOUL.claude.md").is_file()
            or (dir / "propagated" / "SOUL.md").is_file()
            or (
                ((dir / "SOUL.claude.md").is_file() or (dir / "SOUL.md").is_file())
                and has_athlete_band
            )
        ):
            return dir
        # Flat skeleton: soul/ at repo root — not HQ platform/soul/ subtree
        if (
            (dir / "soul").is_dir()
            and not (dir / "engine").is_dir()
            and not (dir / "platform" / "scripts").is_dir()
            and not (dir.parent / "engine" / "soul").is_dir()
            and not (dir.parent / "platform" / "soul").is_dir()
        ):
            return dir
        parent = dir.parent
        if parent == dir:
            break
        dir = parent
    raise RuntimeError(f"Cannot resolve repo root from {file}")


def uses_new_layout(repo: Path) -> bool:
    """True when repo uses user_data/ + gen/ (carved skeleton)."""
    return (repo / "user_data").is_dir()


def is_hq_monorepo(repo: Path) -> bool:
    """HQ monorepo — platform IP + ui; no athlete instance band at root."""
    return (repo / "platform" / "soul").is_dir() and (repo / "ui").is_dir()


def golden_repo_data_dir(repo: Path) -> Path:
    return repo / "shared" / "golden-dataset" / "repo-data"


def engine_root(repo: Path) -> Path:
    eng = repo / "engine"
    return eng if eng.is_dir() else repo


def p(repo: Path, *parts: str) -> Path:
    return engine_root(repo).joinpath(*parts)


def coach_dir(repo: Path) -> Path:
    if uses_new_layout(repo):
        return repo / "user_data" / "coach"
    return repo / "training" / "coach"


def ledger_dir(repo: Path) -> Path:
    if uses_new_layout(repo):
        return repo / "user_data" / "ledger"
    return repo / "training" / "ledger"


def health_dir(repo: Path) -> Path:
    if uses_new_layout(repo):
        return repo / "user_data" / "health"
    return repo / "training" / "health"


def hist_dir(repo: Path) -> Path:
    if uses_new_layout(repo):
        return repo / "user_data" / "activities" / "hist"
    return repo / "training" / "activities" / "history"


def gen_dir(repo: Path) -> Path:
    if uses_new_layout(repo):
        return repo / "gen"
    return repo / "training"


def sessions_dir(repo: Path) -> Path:
    if uses_new_layout(repo):
        return repo / "user_data" / "activities" / "workout_plans" / "sessions"
    return repo / "sessions"


def templates_dir(repo: Path) -> Path:
    if uses_new_layout(repo):
        return repo / "user_data" / "activities" / "workout_plans" / "templates"
    return repo / "templates"


def dashboard_snapshot_path(repo: Path) -> Path:
    if uses_new_layout(repo):
        return repo / "gen" / "dashboard_snapshot.json"
    return repo / "data" / "dashboard_snapshot.json"


def sleep_log_path(repo: Path) -> Path:
    if uses_new_layout(repo):
        return repo / "user_data" / "coach" / "sleep_log.json"
    return repo / "training" / "activities" / "sleep_log.json"


def sync_state_path(repo: Path) -> Path:
    if uses_new_layout(repo):
        return repo / "user_data" / "activities" / "sync_state.json"
    return repo / "training" / "sync_state.json"


def sync_status_path(repo: Path) -> Path:
    if uses_new_layout(repo):
        return repo / "gen" / "sync_status.json"
    return repo / "training" / "sync_status.json"



def quest_history_path(repo: Path) -> Path:
    if uses_new_layout(repo):
        return repo / "gen" / "quest_history.json"
    return repo / "training" / "activities" / "quest_history.json"


def widget_snapshots_path(repo: Path) -> Path:
    if uses_new_layout(repo):
        return repo / "gen" / "widget_snapshots.json"
    return repo / "training" / "widget_snapshots.json"


def chat_history_path(repo: Path) -> Path:
    if uses_new_layout(repo):
        return repo / "user_data" / "coach" / "chat_history.json"
    return repo / "training" / "chat_history.json"


def seasons_dir(repo: Path) -> Path:
    # Grouped with the rest of the coach's archived material (archive/phases.md,
    # archive/week_plans.md), not the live ledger/ - a closed season's challenge_v2.json is a
    # retrospective, not current data.
    if uses_new_layout(repo):
        return repo / "user_data" / "coach" / "archive" / "seasons"
    return repo / "training" / "seasons"


def activities_dir(repo: Path) -> Path:
    if uses_new_layout(repo):
        return repo / "user_data" / "activities"
    return repo / "training" / "activities"


def soul_file_path(repo: Path, target: str = "claude") -> Path:
    """Composed SOUL path for a build target (ADR 0022) - HQ: platform/SOUL.<target>.md;
    athlete repos: propagated/SOUL.claude.md, then repo root.

    The bare `SOUL.md` name is retired at HQ but still probed last in athlete repos:
    `coach-akash` and `coach-skanda` were carved before the split and still carry
    `propagated/SOUL.md`, so dropping the fallback would break the engine scripts running
    inside them.
    """
    if is_hq_monorepo(repo):
        return repo / "platform" / f"SOUL.{target}.md"
    for candidate in (
        repo / "propagated" / f"SOUL.{target}.md",
        repo / "propagated" / "SOUL.md",
        repo / f"SOUL.{target}.md",
    ):
        if candidate.is_file():
            return candidate
    return repo / "SOUL.md"

File: engine/lib/repo-layout.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

/** Walk up from `fromDir` to find repo root (HQ monorepo or athlete skeleton). */
export function repoRoot(fromDir = path.dirname(fileURLToPath(import.meta.url))) {
  let dir = path.resolve(fromDir);
  for (let i = 0; i < 8; i++) {
    // HQ monorepo — platform band + product surfaces
    if (
      fs.existsSync(path.join(dir, "platform", "soul")) &&
      fs.existsSync(path.join(dir, "ui"))
    ) {
      return dir;
    }
    if (fs.existsSync(path.join(dir, "engine", "soul"))) {
      return dir;
    }
    // Athlete repo: a composed SOUL at the propagated/ or repo root. Both the post-ADR-0022
    // `SOUL.claude.md` name and the pre-split `SOUL.md` count — repos carved before the split
    // still carry the old name and must keep resolving.
    const hasAthleteBand =
      fs.existsSync(path.join(dir, "user_data")) || fs.existsSync(path.join(dir, "training"));
    if (
      fs.existsSync(path.join(dir, "propagated", "SOUL.claude.md")) ||
      fs.existsSync(path.join(dir, "propagated", "SOUL.md")) ||
      ((fs.existsSync(path.join(dir, "SOUL.claude.md")) ||
        fs.existsSync(path.join(dir, "SOUL.md"))) &&
        hasAthleteBand)
    ) {
      return dir;
    }
    // Flat athlete skeleton: soul/ at repo root — not HQ's platform/soul/ subtree
    if (
      fs.existsSync(path.join(dir, "soul")) &&
      !fs.existsSync(path.join(dir, "engine")) &&
      !fs.existsSync(path.join(dir, "platform", "scripts")) &&
      !fs.existsSync(path.join(path.dirname(dir), "engine", "soul")) &&
      !fs.existsSync(path.join(path.dirname(dir), "platform", "soul"))
    ) {
      return dir;
    }
    const parent = path.dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  throw new Error(`Cannot resolve repo root from ${fromDir}`);
}

/** true when repo uses user_data/ + gen/ (carved skeleton); false for legacy training/ + data/. */
export function usesNewLayout(repoRootPath) {
  return fs.existsSync(path.join(repoRootPath, "user_data"));
}

/** HQ monorepo — platform IP + ui; no athlete instance band at root. */
export function isHqMonorepo(repoRootPath) {
  return (
    fs.existsSync(path.join(repoRootPath, "platform", "soul")) &&
    fs.existsSync(path.join(repoRootPath, "ui"))
  );
}

export function goldenRepoDataDir(repoRootPath) {
  return path.join(repoRootPath, "shared/golden-dataset/repo-data");
}

export function soulDir(repoRootPath) {
  const inPlatform = path.join(repoRootPath, "platform", "soul");
  if (fs.existsSync(inPlatform)) return inPlatform;
  const inEngine = path.join(repoRootPath, "engine", "soul");
  return fs.existsSync(inEngine) ? inEngine : path.join(repoRootPath, "soul");
}

export function enginePath(repoRootPath, ...parts) {
  const inEngine = path.join(repoRootPath, "engine", ...parts);
  if (fs.existsSync(path.join(repoRootPath, "engine"))) {
    return inEngine;
  }
  return path.join(repoRootPath, ...parts);
}

export function coachDir(repoRootPath) {
  return usesNewLayout(repoRootPath)
    ? path.join(repoRootPath, "user_data", "coach")
    : path.join(repoRootPath, "training", "coach");
}

export function ledgerDir(repoRootPath) {
  return usesNewLayout(repoRootPath)
    ? path.join(repoRootPath, "user_data", "ledger")
    : path.join(repoRootPath, "training", "ledger");
}

export function healthDir(repoRootPath) {
  return usesNewLayout(repoRootPath)
    ? path.join(repoRootPath, "user_data", "health")
    : path.join(repoRootPath, "training", "health");
}

export function histDir(repoRootPath) {
  return usesNewLayout(repoRootPath)
    ? path.join(repoRootPath, "user_data", "activities", "hist")
    : path.join(repoRootPath, "training", "activities", "history");
}

export function genDir(repoRootPath) {
  return usesNewLayout(repoRootPath) ? path.join(repoRootPath, "gen") : path.join(repoRootPath, "training");
}

export function sessionsDir(repoRootPath) {
  return usesNewLayout(repoRootPath)
    ? path.join(repoRootPath, "user_data", "activities", "workout_plans", "sessions")
    : path.join(repoRootPath, "sessions");
}

export function templatesDir(repoRootPath) {
  return usesNewLayout(repoRootPath)
    ? path.join(repoRootPath, "user_data", "activities", "workout_plans", "templates")
    : path.join(repoRootPath, "templates");
}

export function dashboardSnapshotPath(repoRootPath) {
  return usesNewLayout(repoRootPath)
    ? path.join(repoRootPath, "gen", "dashboard_snapshot.json")
    : path.join(repoRootPath, "data", "dashboard_snapshot.json");
}

export function sleepLogPath(repoRootPath) {
  return usesNewLayout(repoRootPath)
    ? path.join(repoRootPath, "user_data", "coach", "sleep_log.json")
    : path.join(repoRootPath, "training", "activities", "sleep_log.json");
}

export function syncStatePath(repoRootPath) {
  return usesNewLayout(repoRootPath)
    ? path.join(repoRootPath, "user_data", "activities", "sync_state.json")
    : path.join(repoRootPath, "training", "sync_state.json");
}

export function syncStatusPath(repoRootPath) {
  return usesNewLayout(repoRootPath)
    ? path.join(repoRootPath, "gen", "sync_status.json")
    : path.join(repoRootPath, "training", "sync_status.json");
}

export function questHistoryPath(repoRootPath) {
  return usesNewLayout(repoRootPath)
    ? path.join(repoRootPath, "gen", "quest_history.json")
    : path.join(repoRootPath, "training", "activities", "quest_history.json");
}

export function widgetSnapshotsPath(repoRootPath) {
  return usesNewLayout(repoRootPath)
    ? path.join(repoRootPath, "gen", "widget_snapshots.json")
    : path.join(repoRootPath, "training", "widget_snapshots.json");
}

export function chatHistoryPath(repoRootPath) {
  return usesNewLayout(repoRootPath)
    ? path.join(repoRootPath, "user_data", "coach", "chat_history.json")
    : path.join(repoRootPath, "training", "chat_history.json");
}

/** Archive seasons - grouped with the rest of the coach's archived material (archive/phases.md,
 * archive/week_plans.md), not the live ledger/ - a closed season's challenge_v2.json is a
 * retrospective, not current data. Old layout uses training/seasons. */
export function seasonsDir(repoRootPath) {
  return usesNewLayout(repoRootPath)
    ? path.join(repoRootPath, "user_data", "coach", "archive", "seasons")
    : path.join(repoRootPath, "training", "seasons");
}

/** Derived plugin snapshots live beside activities in both layouts. */
export function activitiesDir(repoRootPath) {
  return usesNewLayout(repoRootPath)
    ? path.join(repoRootPath, "user_data", "activities")
    : path.join(repoRootPath, "training", "activities");
}

/**
 * Composed SOUL path for a build target (ADR 0022) — HQ: platform/SOUL.<target>.md; athlete
 * repos: propagated/SOUL.claude.md, then repo root.
 *
 * The bare `SOUL.md` name is retired at HQ but still probed last in athlete repos: `coach-akash`
 * and `coach-skanda` were carved before the split and still carry `propagated/SOUL.md`, so
 * dropping the fallback would break the engine scripts running inside them.
 */
export function soulFilePath(repoRootPath, target = "claude") {
  if (isHqMonorepo(repoRootPath)) {
    return path.join(repoRootPath, "platform", `SOUL.${target}.md`);
  }
  const candidates = [
    path.join(repoRootPath, "propagated", `SOUL.${target}.md`),
    path.join(repoRootPath, "propagated", "SOUL.md"),
    path.join(repoRootPath, `SOUL.${target}.md`),
  ];
  for (const candidate of candidates) {
    if (fs.existsSync(candidate)) return candidate;
  }
  return path.join(repoRootPath, "SOUL.md");
}
